[rhel7-branch] Support timezone command usage without timezone specification

M4rtinK opened the pull request:
Make it possible to use the timezone command without a timezone
specification.

This way the timezone command can be used with the --utc flag
to set hardware clock time format while still requiring the user
to select a timezone manually in the Anaconda UI.

Related: rhbz#1312135

M4rtinK commented:
Hope this is not too crazy.

Also the accompanying Anaconda PR is https://github.com/rhinstaller/anaconda/pull/670.

vpodzime commented on pykickstart/commands/timezone.py:
```diff
@@ -35,6 +35,12 @@ def __init__(self, writePriority=0, *args, **kwargs):
+        # - this has been the case up to the F18 version of the command
+        # - this variable has been added to make it possible for later versions
+        #   of the command to drop this requirement
+        self._timezone_spec_required = True
```
This would make more sense to me as a class attribute. But in general, I don't think these changes are worth reusing 3 lines of code in the ``parse()`` method. The RHEL7 version of the command could just have its own ``parse()`` method defined instead.

clumens commented:
In general, I prefer using the inheritance tree instead of variables for specifying whether things are supported in a particular version.  Sometimes that involves being a little wordier, but oh well.  For the RHEL7 version of this patch, I think the RHEL7 class's parse method should handle the possibility of not giving a timezone.  And then for the master version, you'll need to have the same thing in the RHEL7 class and whichever class will be the first in Fedora to support this.

M4rtinK commented:
>In general, I prefer using the inheritance tree instead of variables for specifying whether things are supported in a particular version. Sometimes that involves being a little wordier, but oh well. For the RHEL7 version of this patch, I think the RHEL7 class's parse method should handle the possibility of not giving a timezone. 

Sure - its just that in this case we need to prevent the FC3 parse() method from being called, otherwise we would always get an exception regardless to what the parse() method in the RHEL7 variant does.

So I've changed the PR to implement Vratislavs suggestion - I've just implemented all the stuff the "older" classes do in RHEL7 command class parse() method the way I need it - without calling the ancestors method. 

I've also added a note to the possibly skipped parse() methods so that people don't wonder why they are not being called in RHEL 7.3+.

>And then for the master version, you'll need to have the same thing in the RHEL7 class and whichever class will be the first in Fedora to support this.

Thanks! I wondered how this should be done for the master version - especially with the RHEL7 command variant. I guess it makes sense due to ksvalidator (it needs to be able to validate RHEL7 kickstarts).

clumens commented:
I think this is fine for RHEL7 - having the _getParser method is supposed to help somewhat with code duplication, and if the parse methods are too complicated, we should work to reduce that.  I think the current patch looks good.

clumens commented:
And then for master, you basically have to do two things:  (1) Add this patch to show that RHEL7 now supports it, and (2) Do something very similar adding a new F24 (F25?) class.  But yeah, especially for F24 the parse method is starting to get complex.  At the least, this establishes a point in time where we can start inheriting again.  It's not going to get more complicated.

M4rtinK commented:
I've added a check that throws KickstartParseError when only "timezone" is used in a kickstart & noted this in the documentation.

I've also added unit tests for the new behavior and (hopefully) improved the error messages a bit.

M4rtinK commented:
Pushed; thanks for the review & feedback!
